File: rxbinding/src/main/java/com/jakewharton/rxbinding2/view/ViewGroupHierarchyChangeEvent.java
package com.jakewharton.rxbinding2.view;

import androidx.annotation.NonNull;
import android.view.View;
import android.view.ViewGroup;

public abstract class ViewGroupHierarchyChangeEvent {
  ViewGroupHierarchyChangeEvent() {
  }

  /** The view from which this event occurred. */
  @NonNull public abstract ViewGroup view();

  /** The child from which this event occurred. */
  @NonNull public abstract View child();
}

File: rxbinding/src/main/java/com/jakewharton/rxbinding4/view/ViewGroupHierarchyChangeEvent.kt
package com.jakewharton.rxbinding4.view

import android.view.View
import android.view.ViewGroup

sealed class ViewGroupHierarchyChangeEvent {
  /** The view from which this event occurred.  */
  abstract val view: ViewGroup

  /** The child from which this event occurred.  */
  abstract val child: View
}

/**
 * A child view add event on a [ViewGroup].
 *
 * **Warning:** Instances keep a strong reference to the view. Operators that cache
 * instances have the potential to leak the associated [android.content.Context].
 */
data class ViewGroupHierarchyChildViewAddEvent(
  override val view: ViewGroup,
  override val child: View
) : ViewGroupHierarchyChangeEvent()

/**
 * A child view remove event on a [ViewGroup].
 *
 * **Warning:** Instances keep a strong reference to the view. Operators that cache
 * instances have the potential to leak the associated [android.content.Context].
 */
data class ViewGroupHierarchyChildViewRemoveEvent(
  override val view: ViewGroup,
  override val child: View
) : ViewGroupHierarchyChangeEvent()
